### DISCUSSION TITLE
texequationで段落を切らないようにしたい

### DISCUSSION THREAD

**Discussion by Nyoho** [kmuto/review#1870] [Q&A] [answered]
`//texequation{...}` を使うと、前後で段落が必ず切れてしまいます。

数学では
<img width="494" alt="image" src="https://user-images.githubusercontent.com/118150/206844808-21a7d9ed-4640-423c-8eca-bd972ecfb6a1.png">
のように、文や段落の途中にdisplay数式が入ることはよくあります。
(「を考える」のところが新しい段落になっていない)

これをRe:VIEWでも実現したいです。
Re:VIEWでは `texequation` の前後で必ず段落が途切れてしまいます。

## 実験

サンプル https://github.com/NyohoSampleCodes/review-test-202212 を作って実験しました。

`LATEXBuilder` では `texequation` をほとんどコピーして blank を削除して https://github.com/NyohoSampleCodes/review-test-202212/blob/main/review-ext.rb#L11 と https://github.com/NyohoSampleCodes/review-test-202212/blob/main/review-ext.rb#L41 をしてみましたが、やはりこの前後には空行が入っているみたいで、段落が作られてしまいました。

<img width="589" alt="image" src="https://user-images.githubusercontent.com/118150/206845742-c45b1d76-4968-4f99-803d-cab897acacc1.png">

EPUBを作るときにも全くどこから改造していけば良いのかわかりません。

まずは、EPUBでできれば嬉しいです。ゆくゆくはLATEXBuilderでもできて欲しいです。

どこを改造していけば実現できるでしょうか。

**Comment by kdmsnr** [comment, marked as answer]
~~latexだと `\noindent{}` で囲む必要があるのでしょうか。~~ 

~~であれば、以下のようにするのはどうでしょうか。~~ 

```
--- a/chap1.re
+++ b/chap1.re
@@ -6,4 +6,4 @@
 //texequation{
 0 \to I \to \mathcal{O}_X \to \mathcal{O}_Y \to 0
 //}
-を考える。
+@<embed>{|latex|\\noindent{}を考える。@<embed>{|latex|\}}
```

**Reply by Nyoho** [reply]
@kdmsnr ありがとうございます。

(一旦このスレッドにコピペします。)

```
--- a/chap1.re
+++ b/chap1.re
@@ -6,4 +6,7 @@
 //texequation{
 0 \to I \to \mathcal{O}_X \to \mathcal{O}_Y \to 0
 //}
+//noindent
 を考える。
```

style.css
```
--- a/style.css
+++ b/style.css
@@ -70,6 +70,9 @@ div.lead p {
     line-height: 1.6;
     font-size: 0.75em;
 }
+p.noindent {
+    text-indent: 0;
+}
```

**Reply by Nyoho** [reply]
@kdmsnr ありがとうございます。
この方法ですと、ソースの `texequation` 下すべてに `//noindent` を付けていくということですね。

基本的に見た目がそうなるようにするという解決策ですね。
論理的に段落にする (pの中に入れる) という方法は、あまり現実的ではない……ですかね。^^

この方法でも、もしできるものならば、ソースはそのままで `texequation` の後に空行がなければ自動的にそうなるように、review-ext.rb でなんとかできれば大変ありがたいです。
そんな虫の良い方法はないものでしょうか。

あと、実は段落の最初を話者アイコンで char という独自コマンドを作って 、noindent かつ、

```css
p.noindent {
    text-indent: -2em;
}
```

としてしまっています。
(むしろ話者アイコンのためのこれが筋が悪いのかも知れませんが)

**Reply by kdmsnr** [reply]
review-ext.rb にこういうのを入れるとかでしょうか。

```ruby
module ReVIEW
  module TexEquationOverride
    def texequation(lines, id = nil, caption = '')
      @noindent = false
      texequation_orig(lines, id, caption)
      noindent
    end
  end

  class LATEXBuilder
    alias_method :texequation_orig, :texequation
    prepend TexEquationOverride
  end

  class HTMLBuilder
    alias_method :texequation_orig, :texequation
    prepend TexEquationOverride
  end
end
```

**Reply by kmuto** [reply]
aliasなしで `super(lines, id, caption)` でよさそう?と思いました。

**Reply by kdmsnr** [reply]
確かに……。noindentの初期化も不要ですね。

```ruby
module ReVIEW
  module TexEquationOverride
    def texequation(lines, id = nil, caption = '')
      super(lines, id, caption)
      noindent
    end
  end

  class LATEXBuilder
    prepend TexEquationOverride
  end

  class HTMLBuilder
    prepend TexEquationOverride
  end
end
```

**Reply by Nyoho** [reply]
@kdmsnr さん @kmuto さん、ありがとうございます。この方法でnoindentを自動で入れられました!

そういえば `//image` も入れたときも、段落が改まらないでそのまま続けて欲しいときがあったことも思い出しました。(でもそれは図の位置を考えることでなんとかなりそうです。)

**Comment by kdmsnr** [comment]
あ、すでに `//noindent` コマンドがありますね。

```
--- a/chap1.re
+++ b/chap1.re
@@ -6,4 +6,7 @@
 //texequation{
 0 \to I \to \mathcal{O}_X \to \mathcal{O}_Y \to 0
 //}
+//noindent
 を考える。
```

**Comment by kdmsnr** [comment]
style.css

```
--- a/style.css
+++ b/style.css
@@ -70,6 +70,9 @@ div.lead p {
     line-height: 1.6;
     font-size: 0.75em;
 }
+p.noindent {
+    text-indent: 0;
+}
```

**Comment by kmuto** [comment]
すでにkdmsnrさんから提示があるとおり、基本的には `//noindent` を使うことを推奨です。

Re:VIEW文法で

```
これを

//texequation{
A=B
//}

とする。
```

と

```
これを以下とする。

//texequation{
A=B
//}

次の話に移ろう。
```

を自動判別することは日本語文脈解析の範疇となり、不可能です。

閉じ `//}` 後続の空行有無で挙動を変えるのは、今のパーサの仕組みからすると難しいですし、わかりにくい暗黙ルールを持ち込むことになってしまいます。
「すべてnoindentにする」場合はreview-ext.rbによる挙動変更、一部で使い分けないといけない場合は`//noindent`、という対処になるかと思います。

`inline_m`内で無理矢理判別してゴチャゴチャ処理…ということもちょっと考えましたが、長い式を書くときに辛いですね。

**Reply by Nyoho** [reply]
ありがとうございます。

最初の二つの例で両者を機械的に区別するのが難しいのはよくわかります。

空行がなければ区別できると思いましたが、パーサの仕組みとして難しいんですね。
Re:VIEWでは、普通の地の文であれば、空行があれば段落が改まり、空行がなければ改行があっても段落が続くので、この挙動が `//texequation...` でもできるのではないかなあと、(Re:VIEW構造のことを理解しないまま)無邪気に考えておりました。

**Reply by Nyoho** [reply]
オフトピック気味ですが https://github.com/orangain/textlint-plugin-review という textlint の Re:VIEW用プラグインが、この `texequation` をまたいだ文を1文と取ってくれなくて、まずはHTMLでは <p> 内に入れられないものかと(ずれて)考えたものでした。(きっと textlint とは関係ないですよね)
